### PR TITLE
protocols/rpc: Add getEvents v2 wire types and request validation

### PR DESCRIPTION
**Pull request by urvisavla**
### What

Adds the getEvents v2 wire types to `protocols/rpc`, transcribed from the accepted proposal ([stellar/discussions#1872](https://github.com/orgs/stellar/discussions/1872)): request/response/event structs, scan statuses, error reasons and `error.data` payloads, the spec's caps, and request validation.

The proposal is treated as normative for every field name, enum spelling, error reason, and `invalid_params` message — all are transcribed verbatim and pinned by golden tests.

### Design decisions

- **`Valid(maxFilters MaxFilters)`** checks every rule decidable from the request alone and returns the first violation as a typed `*InvalidParamsError` carrying the proposal's verbatim message plus the ready-to-serialize `error.data` payload. Server-state rules stay with the server: the serving window, unknown-field rejection (needs strict JSON decoding), json-format topic decoding, and the term budget.
- **The three caps take three shapes, matching the proposal's annotations**: the limit cap (`MaxLimitV2 = 1000`) is a flat spec maximum enforced as a constant; the filter cap (256) is a provider-configurable default (`DefaultMaxFiltersV2`) passed to `Valid`; the term budget (15, `DefaultTermBudgetV2`) is provider-configurable and server-enforced — counting terms requires the canonical form of json-format topics, which only the server's XDR-JSON converter can produce (byte-comparing raw JSON would over-count and falsely reject valid queries).
- **Topics are `json.RawMessage`** because their encoding is decided by the sibling `xdrInputFormat` field (base64 string vs XDR-JSON object), so decoding must be deferred to whoever holds the whole request. An explicit JSON `null` topic is normalized to omitted (matches any value) per the proposal.
- **`Limit` is `*uint`**: nil means omitted (the server applies its default of 100); an explicit value outside [1, 1000] is rejected with the spec message, including `"limit": 0`.
- **`"filters": []` is rejected** while an omitted key matches all events — the proposal states "Empty values are invalid: omit `filters` entirely to match all events."

### Notes for the stellar-rpc integration

- Populate `jrpc2.Error.Data` from `InvalidParamsError.Data` — the v1 handler pattern sets only `Code` and `Message`, which would silently drop the proposal's `error.data.reason` contract.
- Enforce the term budget after topic conversion, reporting via `InvalidParamsErrorData`'s `termsUsed`/`termBudget` fields with the proposal's message format.
- Validate the configured filter cap at startup: a zero `maxFilters` rejects any request that sets filters.
- The server owns the default page size (100) applied when `limit` is unset; this package never applies it.

### Testing

Golden JSON assertions for request, response, event, and all three error-data shapes; literal pins for every wire constant, spec cap, and full error message (mutation-tested — value changes fail the suite); a reflection test keeping `EventInfoV2` in lockstep with v1's `EventInfo` minus the deprecated `inSuccessfulContractCall`; validation tables covering every rule including boundary cases for custom and zero filter caps, explicit-null topics, and all four topic positions.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: protocols/rpc/get_events_v2.go [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+
+// GetEventsV2Request is the union of two request shapes: a
+// range query (minLedger, maxLedger, order, filters, xdrInputFormat) or a
+// cursor query (cursor). The two are mutually exclusive; limit and xdrFormat
+// apply to both.
+type GetEventsV2Request struct {
+	MinLedger      uint32          `json:"minLedger,omitempty"`
+	MaxLedger      uint32          `json:"maxLedger,omitempty"`
+	Order          string          `json:"order,omitempty"`
+	Filters        []EventFilterV2 `json:"filters,omitempty"`
```
[the remark this answers is not in the document]
Comment (urvisavla) [reply protocols/rpc/get_events_v2.go:159]: Working as intended, now documented on the field (f809fe4d). Unlike the `limit: 0` case (fixed in the same commit), `"filters": null` has no behavioral surprise: "no filters" and "omitted" both mean match-all, so with a cursor there is no actual conflict to reject either. Rejecting `null` would require presence-tracking custom unmarshalling for a case with identical semantics, and would be inconsistent with the null-topic handling the proposal endorses (an explicit `null` topic position matches any value, same as omitted).

File: protocols/rpc/get_events_v2.go [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+
+// EventFilterV2 is one filter in a range query. Fields within a filter are
+// AND-ed; filters in the list are OR-ed. At least one field must be set.
+// Topic values are ScVals encoded per the request's xdrInputFormat (base64
+// XDR strings, or JSON objects); an omitted topic position matches any
+// value.
+type EventFilterV2 struct {
+	ContractID string          `json:"contractId,omitempty"`
+	EventType  string          `json:"type,omitempty"`
+	Topic0     json.RawMessage `json:"topic0,omitempty"`
```
Comment (tamirms) [inline protocols/rpc/get_events_v2.go:114]: **suggestion:** `RawMessage` is right at the wire layer (topic encoding depends on the sibling `xdrInputFormat`, so no field-level type can decode in isolation — an `xdr.ScValView` field would reject json-format requests at the JSON layer with a Go error). But let's add a constructor so callers don't hand-assemble raw JSON: `encoding/json` encodes any `[]byte`-kinded type without a custom marshaller as a std-base64 string, so for `xdr.ScValView` (which defines no JSON methods) `json.Marshal(view)` already yields the exact wire value — `func TopicScVal(v xdr.ScValView) (json.RawMessage, error)` is a one-liner. Even this PR's tests needed `testTopicB64V2` to build one topic; every downstream client will re-invent that helper, quoting footgun included.

File: protocols/rpc/get_events_v2.go [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+		Message: fmt.Sprintf(format, args...),
+		Data:    InvalidParamsErrorData{Reason: ErrorReasonInvalidParams},
+	}
+}
+
+// Valid checks the request against the proposal's parameter rules and
+// returns the first violation as an *InvalidParamsError. maxFilters is the
+// provider-configured cap on the filters list (DefaultMaxFiltersV2 unless
+// the provider sets its own).
+func (r *GetEventsV2Request) Valid(maxFilters uint) error {
```
Comment (tamirms) [inline protocols/rpc/get_events_v2.go]: **suggestion:** v1's `Valid(maxLimit uint)` and this `Valid(maxFilters uint)` have the same shape but the argument means a different cap. A ported v1 call site — `req.Valid(cfg.MaxEventsLimit)` — compiles cleanly and silently sets the filter cap to 10000. A named type (`type MaxFilters uint`) or a method name carrying the meaning would make the mistake impossible.

Separately, non-blocking: the spec's `"unknown field: contractID"` message is the one wire string this package neither implements nor golden-pins, since it needs strict decoding — which the server has to hand-roll anyway. A strict `Parse` entry point here could own it in a follow-up.

File: protocols/rpc/get_events_v2.go [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+func (r *GetEventsV2Request) validLimit() error {
+	if r.Limit != nil && (*r.Limit < 1 || *r.Limit > MaxLimitV2) {
+		return invalidParamsf("limit must be between 1 and %d", MaxLimitV2)
+	}
+	return nil
+}
+
+func (r *GetEventsV2Request) validFilters(maxFilters uint) error {
+	if r.Cursor != "" {
+		return nil // mutual exclusion is validShape's finding
```
Comment (tamirms) [inline protocols/rpc/get_events_v2.go]: **suggestion:** This duplicated cursor guard is a sign the decomposition doesn't match the spec: the proposal defines `GetEventsRequest = RangeQuery | CursorQuery` discriminated by `cursor`, while `Valid` slices by field category, so each helper re-derives the mode. Branching once on the discriminator removes the guard and reads like the spec:

```go
if r.Cursor != "" { return r.validCursorConflicts() }
return r.validRange(maxFilters) // order, bounds, xdrInputFormat, filters
```

Bonus: `cursor` + invalid `xdrInputFormat` then reports the mutual-exclusion error instead of "xdrInputFormat must be…" — the truer complaint, since `xdrInputFormat` isn't a legal cursor-query field at all.

File: protocols/rpc/get_events_v2.go [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+	}
+	if r.Filters != nil && (len(r.Filters) == 0 || uint(len(r.Filters)) > maxFilters) {
+		return invalidParamsf("filters must contain 1 to %d filters", maxFilters)
```
Comment (tamirms) [inline protocols/rpc/get_events_v2.go:244]: **nit:** A misconfigured `maxFilters` of 0 emits "filters must contain 1 to 0 filters". Either note in `Valid`'s doc that the cap must be ≥ 1 or special-case zero.

File: protocols/rpc/get_events_v2.go [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+		if err := r.Filters[i].valid(i, r.XDRInputFormat); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (f *EventFilterV2) valid(index int, xdrInputFormat string) error {
+	topics := f.Topics()
+	hasTopic := false
```
Comment (tamirms) [inline protocols/rpc/get_events_v2.go]: **nit:**
```go
hasTopic := slices.ContainsFunc(topics[:], func(t json.RawMessage) bool { return t != nil })
```

File: protocols/rpc/get_events_v2.go [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+	// XDR-JSON converter and is validated there. An invalid format is
+	// validFormats's finding, so no topic error is derived from it.
+	if xdrInputFormat != "" && xdrInputFormat != FormatBase64 {
+		return nil
+	}
+	for pos, topic := range topics {
+		if topic == nil {
+			continue
+		}
+		var b64 string
```
Comment (tamirms) [inline protocols/rpc/get_events_v2.go]: **suggestion:** This decodes each topic into a full `xdr.ScVal` tree and discards it. Unmarshalling into a view validates without allocating:

```go
var view xdr.ScValView
if err := json.Unmarshal(topic, &view); err != nil { // base64-decodes for free
    return invalidParamsf(...)
}
raw, err := view.Raw()
if err != nil || len(raw) != len(view) { // reject malformed + trailing bytes
    return invalidParamsf(...)
}
```

`Raw()` runs full structural validation; the length check keeps `SafeUnmarshalBase64`'s trailing-garbage strictness. Same canonical bytes the server can later match with `bytes.Equal` against event-topic views.

File: protocols/rpc/get_events_v2.go [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+// TermBudget are present only when the query exceeds the term budget.
+type InvalidParamsErrorData struct {
+	Reason     string `json:"reason"`
+	TermsUsed  uint32 `json:"termsUsed,omitempty"`
+	TermBudget uint32 `json:"termBudget,omitempty"`
+}
+
+// LedgerOutOfRangeErrorData is error.data for ledger_out_of_range.
+type LedgerOutOfRangeErrorData struct {
+	Reason        string `json:"reason"`
```
Comment (tamirms) [inline protocols/rpc/get_events_v2.go:79]: **suggestion:** The type name already encodes the reason, but the field is caller-filled — nothing prevents `LedgerOutOfRangeErrorData{Reason: "cursor_malformed"}`. A trivial `MarshalJSON` injecting it (or constructors) makes the illegal state unrepresentable. (Same for `CursorMalformedErrorData` below.) `InvalidParamsErrorData` is the exception since the server reuses it for term-budget errors — worth a comment saying so.

File: protocols/rpc/get_events_v2.go [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+	// DefaultTermBudgetV2 is the default cap on distinct index terms per
+	// query. Counting terms needs the canonical form of json-format topics,
+	// which only the server's XDR-JSON converter can produce, hence the
+	// server enforces it (reporting via InvalidParamsErrorData's TermsUsed
+	// and TermBudget).
+	DefaultTermBudgetV2 = 15
+)
+
+const (
+	OrderAscending  = "asc"
```
Comment (tamirms) [inline protocols/rpc/get_events_v2.go:40]: **nit, take or leave:** These closed vocabularies (and the scan statuses below) could be defined string types (`type Order string`, `type ScanStatus string`) — identical on the wire, self-documenting in the struct fields. Plain strings do match the package's existing style, so fine either way.

File: protocols/rpc/get_events_v2_test.go [hunk 10]
```diff
@@ -0,0 +1,531 @@
...
+		},
+		{
+			name:    "invalid xdrInputFormat",
+			request: GetEventsV2Request{MinLedger: 1, XDRInputFormat: "hex"},
+			wantErr: `xdrInputFormat must be "base64" or "json"`,
+		},
+	})
+}
+
+func TestGetEventsV2RequestValidCursor(t *testing.T) {
```
Comment (tamirms) [inline protocols/rpc/get_events_v2_test.go:196]: **test gap:** The doc comment on `Filters` promises explicit `"filters": null` decodes like an omitted member, but no test unmarshals real JSON to pin it — worth one case each for `{"cursor":"x","filters":null}` (valid) and `{"minLedger":1,"filters":null}` (valid, matches all).

File: protocols/rpc/get_events_v2_test.go [hunk 11]
```diff
@@ -0,0 +1,531 @@
...
+	for i := range typ.NumField() {
+		field := typ.Field(i)
+		tag := field.Tag.Get("json")
+		require.NotEmpty(t, tag)
+		specs[strings.Split(tag, ",")[0]] = tag + " " + field.Type.String()
+	}
+	return specs
+}
+
+func mapKeys(m map[string]json.RawMessage) []string {
```
Comment (tamirms) [inline protocols/rpc/get_events_v2_test.go]: **nit:** On go 1.25: `slices.Collect(maps.Keys(m))`.

**Review by tamirms — COMMENTED** [review]
A few suggestions inline, nothing blocking. One non-comment note: the `EventInfoV2`/`EventInfo` duplication with the reflection mirror test is the right call — a shared embedded struct would reorder v1's JSON keys and break its construction API. Nice test.

File: protocols/rpc/get_events_v2.go (hunk 2 above)
Comment (urvisavla) [reply protocols/rpc/get_events_v2.go:114]: Done in 6650d9cb: `TopicScVal(xdr.ScValView) (json.RawMessage, error)` — it validates the view first (`Raw()` + trailing-bytes check) so garbage errors at construction, then marshals. `testTopicB64V2` now goes through it.

File: protocols/rpc/get_events_v2.go (hunk 3 above)
Comment (urvisavla) [reply protocols/rpc/get_events_v2.go]: Done in 6650d9cb: `Valid` now takes a defined `type MaxFilters uint`, so the ported v1 call site no longer compiles. Agreed on the strict `Parse` entry point owning the unknown-field message — deferring that to a follow-up as suggested.

File: protocols/rpc/get_events_v2.go (hunk 4 above)
Comment (urvisavla) [reply protocols/rpc/get_events_v2.go]: Done in 6650d9cb: `Valid` branches once on the cursor discriminator into `validCursorQuery`/`validRangeQuery`; the duplicated guard is gone, and the bonus behavior is real and pinned — `cursor` + invalid `xdrInputFormat` now reports mutual exclusion (new test case).

File: protocols/rpc/get_events_v2.go (hunk 5 above)
Comment (urvisavla) [reply protocols/rpc/get_events_v2.go:244]: Took the doc-note option (6650d9cb): `Valid` documents that maxFilters must be at least 1, and the zero-cap behavior is pinned by a test. Chose not to special-case zero since any in-`Valid` handling either invents a non-spec client message or silently substitutes a default — config validation belongs at server startup.

File: protocols/rpc/get_events_v2.go (hunk 6 above)
Comment (urvisavla) [reply protocols/rpc/get_events_v2.go]: Done in 6650d9cb.

File: protocols/rpc/get_events_v2.go (hunk 7 above)
Comment (urvisavla) [reply protocols/rpc/get_events_v2.go]: Done in 6650d9cb: topics now unmarshal into `xdr.ScValView` with `Raw()` + length check for trailing-byte strictness — no ScVal tree allocated.

File: protocols/rpc/get_events_v2.go (hunk 8 above)
Comment (urvisavla) [reply protocols/rpc/get_events_v2.go:79]: Done in 6650d9cb: both structs inject Reason in `MarshalJSON` (the golden tests now construct them without Reason to prove it), and `InvalidParamsErrorData` carries a comment explaining why it is the exception.

File: protocols/rpc/get_events_v2.go (hunk 9 above)
Comment (urvisavla) [reply protocols/rpc/get_events_v2.go:40]: Leaving as plain strings per the take-or-leave — matching the package's existing style (`EventType*`, `Format*` are plain strings too).

File: protocols/rpc/get_events_v2_test.go (hunk 10 above)
Comment (urvisavla) [reply protocols/rpc/get_events_v2_test.go:196]: Done in 6650d9cb: `TestGetEventsV2FiltersNullJSON` unmarshals real JSON for both shapes — `{"cursor":"x","filters":null}` and `{"minLedger":1,"filters":null}` — and asserts both validate.

File: protocols/rpc/get_events_v2_test.go (hunk 11 above)
Comment (urvisavla) [reply protocols/rpc/get_events_v2_test.go]: Done in 6650d9cb — helper deleted in favor of `slices.Collect(maps.Keys(m))`.

**Review by tamirms — APPROVED** [review]
